storybookjs/storybook
# Angular: Remove cached NgModules and introduce a global queue during bootstrapping

Closes #24813






## What I did



Problem: Caching the NgModule per component, causes the NgModule generated for the first story to get used for the following stories, but they may not have generated the same NgModule.

### Explanation:

Assume the following is the CSF file:
```ts
export default {
  component: FooComponent,
}

export const Primary = { }
export const Secondary = {
  moduleMetadata: {
    declarations: [ BarComponent ],
  },
}
```

In the renderer, the following is basically what the NgModule would be for each story:
```ts
// Primary
@NgModule({
  declarations: [ FooComponent ],
})
class StorybookModule {}

// Secondary
@NgModule({
  declarations: [ FooComponent, BarComponent ],
})
class StorybookModule {}
```

When using inline rendering to render both stories in a single document, Storybook attempts to render each story, without waiting on the previous to complete. The problem is that Angular only allows a component to be declared by a single NgModule, so we have to clear the compiled component cache after each story's app is initialized.

The solution being used currently, would cache the NgModule by the `component` defined in `Meta`. That avoids the error, but breaks the "Secondary" story, because `BarComponent` is not declared in the NgModule that was cached when the "Primary" story rendered (*Assuming the "Primary" story rendered first.*).

I don't know if the way inline stories render changed and used to wait for completion of each story or if it just wasn't fully implemented as expected. `AbstractRenderer` already has a method that calls `ɵresetCompiledComponents` and both `CanvasRenderer` and `DocsRenderer` call it after a full render. The problem is two render methods can be running in parallel, which makes the reset not work correctly.

To fix `ɵresetCompiledComponents` not getting called correctly, a queue is introduced to the `AbstractRenderer` that prevents calling `bootstrapApplication` in parallel. This may have an effect on performance, if a particular story is slow to initialize, but I don't know of another way to avoid conflicts between each Angular application's compiling.

## Checklist for Contributors

### Testing



#### The changes in this PR are covered in the following automated tests:

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

To see the problem with caching the NgModule, 
1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template angular-cli/default-ts`
2. In the sandbox file `stories/core/decorators/componentWrapperDecorator/decorators.stories.ts`, add `tags: ['autodocs']` to the default export `Meta`.
3. Open Storybook in your browser
4. Access "frameworks > angular > core > decorators > componentWrapperDecorator > decorators > Docs" stories. Before this change, all the stories render something, but any of the stories declaring `ParentComponent` threw an error, because the first story did not declare `ParentComponent`. With this change all should render correctly.

To see the reason for adding the lock, instead of just removing the NgModule caching, you can do the same steps, but comment out `await acquireLock(selector);` in the the `bootstrapLock` function in `code/frameworks/angular/src/client/angular-beta/utils/BootstrapLock.ts`. Commenting that line would disable the lock and cause the *multiple modules declaring component* error that the caching was supposed to fix.

### Documentation



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release



This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_
